File: extractor-app/src/main/java/ru/mydesignstudio/database/metadata/extractor/runner/ApplicationRunner.java
package ru.mydesignstudio.database.metadata.extractor.runner;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.extractors.DatabaseMetadataExtractor;
import ru.mydesignstudio.database.metadata.extractor.extractors.TableMetadataExtractor;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.DatabaseMetadata;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.output.MetadataOutput;
import ru.mydesignstudio.database.metadata.extractor.parameters.Parameters;
import ru.mydesignstudio.database.metadata.extractor.parameters.ParametersReader;

import java.util.List;

@Slf4j
@Component
public class ApplicationRunner implements CommandLineRunner {
  @Autowired
  private TableMetadataExtractor tableExtractor;

  @Autowired
  private DatabaseMetadataExtractor databaseExtractor;

  @Autowired
  private MetadataOutput metadataOutput;

  @Autowired
  private ParametersReader parametersReader;

  @Override
  public void run(String... args) throws Exception {
    log.info("Reading parameters file");
    final Parameters parameters = parametersReader.read();
    log.info("Done");

    log.info("Reading database metadata");
    final List<DatabaseMetadata> databaseMetadata = databaseExtractor.extract(parameters.getSources());
    log.info("Done");

    log.info("Reading tables metadata");
    final List<TableMetadata> tableMetadata = tableExtractor.extract(parameters.getSources());
    log.info("Done");

    log.info("Output");
    metadataOutput.output(databaseMetadata, tableMetadata);
    log.info("Done");
  }
}

File: extractor-app/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/runner/ApplicationRunner.kt
package ru.mydesignstudio.database.metadata.extractor.runner

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.CommandLineRunner
import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.reader.ParametersReader
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.validator.ParametersValidator
import ru.mydesignstudio.database.metadata.extractor.registry.SourceMetadataExtractorRegistry
import java.util.stream.Collectors

@Component
class ApplicationRunner @Autowired constructor (
        val sourceRegistry: SourceMetadataExtractorRegistry,
        val parametersReader: ParametersReader,
        val parametersValidator: ParametersValidator
        ) : CommandLineRunner {

    private val logger = LoggerFactory.getLogger(ApplicationRunner::class.java)

    override fun run(vararg args: String) {
        logger.info("Reading parameters file")
        val parameters = parametersReader.read()
        logger.info("Done")

        logger.info("Validating given parameters")
        val validationResult = parametersValidator.validate(parameters)
        if (!validationResult.isValid) {
            throw RuntimeException("Parameters are invalid: ${validationResult.message}")
        }
        logger.info("Parameters are correct")

        logger.info("Extracting metadata from sources")
        var metadata = parameters.sources.stream()
                .map { Pair(it, sourceRegistry.getExtractor(it.source)) }
                .map { p -> p.second.extract(p.first) }
                .flatMap { it.stream() }
                .collect(Collectors.toList())

        /*
        logger.info("Reading database metadata")
        val databaseMetadata = databaseExtractor.extract(parameters.sources)
        logger.info("Done")

        logger.info("Reading tables metadata")
        val tableMetadata = tableExtractor.extract(parameters.sources)
        logger.info("Done")

        logger.info("Output")
        metadataOutput.output(databaseMetadata, tableMetadata)
        logger.info("Done")
         */
    }
}